### ISSUE TITLE
scatter_add/scatter_reduce lowers to O(index_len / 256) serially-chained ttnn.scatter ops - linear blowup makes GNN aggregation intractable

### ISSUE THREAD

**Issue by St1viG** [tenstorrent/tt-mlir#8714]
### Describe the bug

Filing here as a lowering issue; please redirect to tt-xla if more appropriate.

A single logical `scatter_add` / `scatter_reduce` over a length-L index is lowered to a **serial chain of `ttnn.scatter` ops, one per 256-element slice** of the index/source, each writing the full destination buffer and feeding the next. The number of emitted `ttnn.scatter` ops therefore grows **linearly with the index size** (≈ L_elems / 128 including the fwd+bwd pair). The op compiles and produces correct results — this is a **lowering efficiency** issue, not a correctness or unsupported-op issue.

For GNN message passing (PyG `MessagePassing` uses `scatter_add`/`scatter_reduce_` for aggregation) this is fatal: full PubMed (88,648 edges, 64 features) produces ~190k chained scatter ops and a single forward pass does not finish compiling in 38+ minutes.

## Observed scaling (N=512, F=8, sweeping E)
| E (index len) | `ttnn.scatter` ops |
|--------------:|-------------------:|
|           256 |                 16 |
|         1,024 |                 64 |
|         4,096 |                256 |
|        16,384 |              1,024 |

Linear — 16 ops per 256-element slice region. Each emitted op:
"ttnn.scatter"(%prev, %idx256, %src256)
   <{dim = 0, scatter_reduce_type = #ttcore.reduce_type}>
   : (tensor, tensor<256xsi32>, tensor<256xf32>) -> tensor
with `ttnn.slice_static` ops carving the index/src into consecutive 256-wide chunks.

## Impact
Blocks end-to-end GNN training (GATv2/GCN/GraphSAGE) on real-sized graphs on TT. The serial chaining (each op depends on the previous) also serializes execution, compounding the cost.

### Steps to reproduce the issue

1. Save the script below as repro.py:

import os
os.environ.setdefault("TTXLA_LOGGER_LEVEL", "DEBUG")
import torch, torch_xla
import torch_xla.runtime as xr
xr.set_device_type("TT"); os.environ["XLA_STABLEHLO_COMPILE"] = "1"

N, E, F = 512, 4096, 8          # nodes, edges (index len), features
dev = torch_xla.device()
index = torch.randint(0, N, (E,), dtype=torch.long, device=dev)
src   = torch.randn(E, F, device=dev)
out   = torch.zeros(N, F, device=dev)
out = out.scatter_add(0, index.unsqueeze(-1).expand(-1, F), src)
torch_xla.sync(wait=True)

2. Count the emitted scatter ops:

TTXLA_LOGGER_LEVEL=DEBUG python repro.py 2>&1 | grep -c "ttnn.scatter"

3. Re-run with increasing E (256, 1024, 4096, 16384). The ttnn.scatter count grows linearly with E.

### Expected behavior

One logical scatter / segment-reduce should lower to a single (or O(1)) device op over the whole index — e.g. a fused segment-sum — rather than O(index_len / 256) serially-chained ttnn.scatter ops.

### Please complete the following environment information

- pjrt-plugin-tt: 1.2.0.dev20260527003051
- torch_xla: 2.9.0+git5c82b10
- torch: 2.10.0
- Hardware: Wormhole b0 (N300)
- OS: Ubuntu 24.04, TT-KMD 2.8.0

**Comment by sdjordjevicTT** [comment]
Hi @St1viG, thanks for bringing this to our attention! Unfortunately, this issue stems from limitations in the metal operation implementation that our compiler is currently targeting. We will need to reassess the current state of this operation in tt-metal and see whether we can remove some restrictions from the compiler code.

**Comment by sdjordjevicTT** [comment]
I created the following PR to address this issue:
https://github.com/tenstorrent/tt-mlir/pull/8718

We'll need some time to run the frontend tests before committing, but we should merge the change during the next week.

**Comment by sdjordjevicTT** [comment]
Hi @St1viG, I just merged the fix into the compiler. Please note that it will take some time to bubble up to the tt-xla.
